cpp/src/arrow/chunk_resolver.h

```diff
@@ -36,8 +39,42 @@ struct ChunkLocation {
 
   /// \brief Index of the value in the chunk
   ///
-  /// The value is undefined if chunk_index >= chunks.size()
+  /// The value is UNDEFINED if chunk_index >= chunks.size()
   int64_t index_in_chunk = 0;
+
+  /// \brief Create a ChunkLocation without asserting any preconditions.
+  static ChunkLocation Forge(int64_t chunk_index, int64_t index_in_chunk) {
```

Comment:
Can we avoid inventing terminology that's currently not used in the project?

Reply:
Is there an actual need for a checked constructor?

Reply:
It adds a layer of safety and by poisoning the `index_in_chunk` value when `chunk_index` is invalid, I (and other users of the library) are more likely to detect unguarded use of `chunk_index`.

Reply:
I don't understand how such unguarded uses could occur. The primary use case for `ChunkLocation` is as a return value for `ChunkResolver`. _Unless_ the user has passed an invalid logical index, `ChunkResolver` should always return valid `ChunkLocation`s.